src/NServiceBus.AmazonSQS/MessagePump.cs

```diff
@@ -61,220 +60,256 @@ public MessagePump(ConnectionConfiguration configuration, IAmazonS3 s3Client, IA
+        }
+
+        async Task ConsumeMessages(CancellationToken token)
+        {
+            // cached and reused per receive loop
+            var concurrentReceiveOperations = new List<Task>(numberOfMessagesToFetch);
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    var receivedMessages = await sqsClient.ReceiveMessageAsync(receiveMessagesRequest, token).ConfigureAwait(false);
+
+                    ProcessMessages(receivedMessages.Messages, concurrentReceiveOperations, token);
+
+                    await Task.WhenAll(concurrentReceiveOperations).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // ignore for graceful shutdown
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error("Exception thrown when consuming messages", ex);
+                }
+                finally
+                {
+                    concurrentReceiveOperations.Clear();
+                }
+            } // while
+        }
+
+        // ReSharper disable once ParameterTypeCanBeEnumerable.Local
+        // ReSharper disable once SuggestBaseTypeForParameter
+        void ProcessMessages(List<Message> receivedMessages, List<Task> concurrentReceiveOperations, CancellationToken token)
+        {
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var receivedMessage in receivedMessages)
+            {
+                concurrentReceiveOperations.Add(ProcessMessage(receivedMessage, token));
+            }
+        }
+
+        async Task ProcessMessage(Message receivedMessage, CancellationToken token)
+        {
             try
             {
-                await Task.WhenAll(consumerTasks.ToArray()).ConfigureAwait(false);
+                await maxConcurrencySempahore.WaitAsync(token).ConfigureAwait(false);
             }
             catch (OperationCanceledException)
             {
-                // Silently swallow OperationCanceledException
-                // These are expected to be thrown when _cancellationTokenSource
-                // is Cancel()ed above.
+                // shutting, semaphore doesn't need to be released because it was never acquired
+                return;
             }
 
-            cancellationTokenSource?.Dispose();
-            maxConcurrencySempahore?.Dispose();
+            try
+            {
+                IncomingMessage incomingMessage = null;
+                TransportMessage transportMessage = null;
+
+                var isPoisonMessage = false;
+                try
+                {
+                    transportMessage = JsonConvert.DeserializeObject<TransportMessage>(receivedMessage.Body);
+
+                    incomingMessage = await transportMessage.ToIncomingMessage(s3Client, configuration, token).ConfigureAwait(false);
```

Comment:
I've raised a separate issued https://github.com/Particular/NServiceBus.AmazonSQS/issues/176 to deal with this.